Wrap and submit the exported buffer forward instead of blitting

elima opened the pull request:
Since we render the exported buffer to the full size of the surface, we don't need to blit it.

This patch creates a wl_buffer that wraps the buffer resource and attach it to the wayland compositor surface for rendering, thus avoiding texture mapping and GLES altogether.

Tested in Intel/i965 and i.MX6 platforms, in both cases with a noticeable increase in performance.

aperezdc commented on platform/cog-platform-fdo.c:
```diff
@@ -998,6 +987,11 @@ on_surface_frame (void *data, struct wl_callback *callback, uint32_t time)
+    if (wpe_view_data.buffer != NULL) {
+        wl_buffer_destroy (wpe_view_data.buffer);
+        wpe_view_data.buffer = NULL;
+    }
```
Please replace the whole `if`-statement with [g_clear_pointer()](https://developer.gnome.org/glib/unstable/glib-Memory-Allocation.html#g-clear-pointer), which is equivalent:

```c
g_clear_pointer (&wpe_view_data.buffer, wl_buffer_destroy);
```

Most of the code in Cog uses the `g_clear_*` and `g_auto*` macros, and I think once one gets used to them, the code is more concise without being less explicit or unclear.

aperezdc replied:
Please replace this whole `if`-statement with [g_clear_pointer()](https://developer.gnome.org/glib/unstable/glib-Memory-Allocation.html#g-clear-pointer), which is equivalent:

```c
g_clear_pointer (&wpe_view_data.buffer, wl_buffer_destroy);
```

Most of the code in Cog uses the `g_clear_*` and `g_auto*` macros, and I think once one gets used to them, the code is more concise without being less explicit or unclear.

aperezdc commented on platform/cog-platform-fdo.c:
```diff
@@ -1069,7 +1026,26 @@ on_export_buffer_resource (void* data, struct wl_resource* buffer_resource)
+    if (eglCreateWaylandBufferFromImageWL == NULL) {
+        eglCreateWaylandBufferFromImageWL =
+            eglGetProcAddress ("eglCreateWaylandBufferFromImageWL");
+        assert (eglCreateWaylandBufferFromImageWL != NULL);
```
Please replace the assertion with:

```c
g_assert_nonnull (eglCreateWaylandBufferFromImageWL);
```

Using `g_assert*` gets us better messages when the assertion checks fail.

aperezdc commented on platform/cog-platform-fdo.c:
```diff
@@ -1069,7 +1026,26 @@ on_export_buffer_resource (void* data, struct wl_resource* buffer_resource)
+
+    wpe_view_data.buffer = eglCreateWaylandBufferFromImageWL (egl_data.display,
+                                                              wpe_view_data.image);
+    assert (wpe_view_data.buffer != NULL);
```
Ditto, please use `g_assert_nonnull()`.

elima commented:
@aperezdc, thanks for the review. I have addressed all your comments and pushed a new version.

Eventually I will have to sit down and glib-ify and organize platform-fdo.

aperezdc commented:
@elima Thanks for updating the PR, it's merged now.

BTW, no hurries in cranking up the GLib usage, the code is working pretty well as it is. We can always slowly introduce things like in this patch when possible :+1:
